cvmfs/magic_xattr.cc

```diff
@@ -0,0 +1,569 @@
+  base_xattrs_["user.repo_metainfo"] = new RepoMetainfoMagicXattr();
+  base_xattrs_["user.revision"] = new RevisionMagicXattr();
+  base_xattrs_["user.root_hash"] = new RootHashMagicXattr();
+  base_xattrs_["user.rx"] = new RxMagicXattr();
+  base_xattrs_["user.speed"] = new SpeedMagicXattr();
+  base_xattrs_["user.tag"] = new TagMagicXattr();
+  base_xattrs_["user.timeout"] = new TimeoutMagicXattr();
+  base_xattrs_["user.timeout_direct"] = new TimeoutDirectMagicXattr();
+  base_xattrs_["user.usedfd"] = new UsedFdMagicXattr();
+  base_xattrs_["user.useddirp"] = new UsedDirPMagicXattr();
+  base_xattrs_["user.version"] = new VersionMagicXattr();
+
+  withhash_xattrs_["user.hash"] = new HashMagicXattr();
+  withhash_xattrs_["user.lhash"] = new LHashMagicXattr();
+
+  regular_xattrs_["user.chunk_list"] = new ChunkListMagicXattr();
+  regular_xattrs_["user.chunks"] = new ChunksMagicXattr();
+  regular_xattrs_["user.compression"] = new CompressionMagicXattr();
+  regular_xattrs_["user.external_file"] = new ExternalFileMagicXattr();
+
+  symlink_xattrs_["user.rawlink"] = new RawlinkMagicXattr();
+  symlink_xattrs_["xfsroot.rawlink"] = symlink_xattrs_["user.rawlink"];
+
+  authz_xattrs_["user.authz"] = new AuthZMagicXattr();
+}
+
+std::string MagicXattrManager::GetListString(catalog::DirectoryEntry *dirent) {
+  std::string result;
+  std::map<std::string, BaseMagicXattr *>::iterator it = base_xattrs_.begin();
+  for (; it != base_xattrs_.end(); ++it) {
+    result += (*it).first + std::string("\0", 1);
+  }
+  if (!dirent->checksum().IsNull()) {
+    it = withhash_xattrs_.begin();
+    for (; it != withhash_xattrs_.end(); ++it) {
+      result += (*it).first + std::string("\0", 1);
+    }
+  }
+  if (dirent->IsRegular()) {
+    it = regular_xattrs_.begin();
+    for (; it != regular_xattrs_.end(); ++it) {
+      result += (*it).first + std::string("\0", 1);
+    }
+  }
+  if (dirent->IsLink()) {
+    it = symlink_xattrs_.begin();
+    for (; it != symlink_xattrs_.end(); ++it) {
+      result += (*it).first + std::string("\0", 1);
+    }
+  }
+  if (mount_point_->has_membership_req()) {
+    it = authz_xattrs_.begin();
+    for (; it != authz_xattrs_.end(); ++it) {
+      result += (*it).first + std::string("\0", 1);
+    }
+  }
+  return result;
+}
+
+BaseMagicXattr *MagicXattrManager::Get(const std::string &name,
+                                       PathString path,
+                                       catalog::DirectoryEntry *d)
+{
+  BaseMagicXattr *result;
+  if (base_xattrs_.count(name) > 0) {
+    result = base_xattrs_[name];
+  } else if (withhash_xattrs_.count(name) > 0) {
+    result = withhash_xattrs_[name];
+  } else if (regular_xattrs_.count(name) > 0) {
+    result = regular_xattrs_[name];
+  } else if (symlink_xattrs_.count(name) > 0) {
+    result = symlink_xattrs_[name];
+  } else if (authz_xattrs_.count(name) > 0) {
+    result = authz_xattrs_[name];
+  } else {
+    return NULL;
+  }
+
+  result->path_ = path;
```

Comment:
Hm, here we have a thread-safety issue.  For the mountpoint it's easy, it is fixed for the lifetime of the manager, so in fact the magic xattr objects can get the mountoint on construction.

Path and dirent need to be locked somehow.  I think the magic xattr base class can have `Lock(path, dirent)` and `Release()` methods that make sure that path and dirent stay stable for `PrepareValueFenced()` followed by `Get()`.

The `MagicXattrManager::Get()` can then return an RAII wrapper around the xattr object that calls `Release()` on destruction.